File: skillnet-frontend/src/app/core/services/notification.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface AppNotification {
  id: number;
  notificationType?: string;
  title: string;
  message: string;
  read: boolean;
  createdAt?: string;
  link?: string;
}

export interface NotificationCount {
  unreadCount: number;
}

@Injectable({ providedIn: 'root' })
export class NotificationService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/notifications`;

  list(): Observable<AppNotification[]> {
    return this.http.get<AppNotification[]>(this.baseUrl);
  }

  count(): Observable<NotificationCount> {
    return this.http.get<NotificationCount>(`${this.baseUrl}/count`);
  }

  markRead(id: number): Observable<AppNotification> {
    return this.http.post<AppNotification>(`${this.baseUrl}/${id}/read`, null);
  }

  markAllRead(): Observable<{ updated: number }> {
    return this.http.post<{ updated: number }>(`${this.baseUrl}/read-all`, null);
  }
}

File: backend/src/main/java/com/skillnet/service/notification/NotificationService.java
package com.skillnet.service.notification;

import com.skillnet.persistence.entity.core.Notification;
import com.skillnet.persistence.repository.NotificationRepository;
import com.skillnet.web.dto.response.NotificationCountResponseDTO;
import com.skillnet.web.dto.response.NotificationResponseDTO;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
@RequiredArgsConstructor
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @Transactional(readOnly = true)
    public List<NotificationResponseDTO> listForUser(Long userId) {
        return notificationRepository.findByUser_IdOrderByCreatedAtDesc(userId).stream()
                .map(this::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public NotificationCountResponseDTO countUnread(Long userId) {
        long count = notificationRepository.countByUser_IdAndReadFalse(userId);
        return NotificationCountResponseDTO.builder().unreadCount(count).build();
    }

    @Transactional
    public NotificationResponseDTO markRead(Long userId, Long notificationId) {
        Notification notification = notificationRepository
                .findById(notificationId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Notificación no encontrada"));

        if (!notification.getUser().getId().equals(userId)) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "No autorizado");
        }

        notification.setRead(true);
        return toResponse(notificationRepository.save(notification));
    }

    @Transactional
    public int markAllRead(Long userId) {
        List<Notification> notifications = notificationRepository.findByUser_IdOrderByCreatedAtDesc(userId);
        int updated = 0;
        for (Notification notification : notifications) {
            if (!notification.isRead()) {
                notification.setRead(true);
                updated++;
            }
        }
        notificationRepository.saveAll(notifications);
        return updated;
    }

    private NotificationResponseDTO toResponse(Notification notification) {
        return NotificationResponseDTO.builder()
                .id(notification.getId())
                .notificationType(notification.getNotificationType())
                .title(notification.getTitle())
                .message(notification.getMessage())
                .read(notification.isRead())
                .createdAt(notification.getCreatedAt())
                .link(notification.getLink())
                .build();
    }
}
